### DISCUSSION TITLE
What would be the best way to distribute a binary plugin

### DISCUSSION THREAD

**Discussion by ModProg** [wbthomason/packer.nvim#1107] [Q&A] [answered]
I have a plugin written in rust: https://github.com/ModProg/config.nvim, and while there is a way to use it by just having it compile the plugin on update through `cargo build` that requires the necessary build dependencies and the additional time it takes to compile.

I was wondering what would be the best way to install my plugin binarry directly (I have them attached to each release: https://github.com/ModProg/config.nvim/releases/tag/v0.0.4).
They are zip files which contain the exact directory structure needed to run the plugin, i.e. this zip should be extracted directly into the installation folder.

I know I could just write a bash script to trigger installation, but I was wondering if there might be a better way to do his.

**Comment by yutkat** [comment]
I know this plugin is made by Rust, but this plugin is installed by shellscript.
https://github.com/shaunsingh/oxocarbon.nvim

**Reply by ModProg** [reply]
Well I have implemented something similar:
plugins.lua:
```lua
{
    'ModProg/test.nvim',
    run = function(plugin)
        local await = require 'packer.async'.wait
        local jobs = require 'packer.jobs'

        local neovim_version = vim.version().major .. '.' .. vim.version().minor
        local os
        if jit.os == "Linux" then
            os = "linux"
        elseif jit.os == "Windows" then
            os = "windows"
        else
            os = "macos"
        end
        await(jobs.run({ "./install", "download", os, neovim_version, "latest" }, { cwd = plugin.install_path }))
    end,
}
```
install
```bash
#!/bin/sh

crate_name=test
github_repo=ModProg/test.nvim
module_name=test

installation_type=$1
os=$2
neovim_version=$3
version=$4

if [ $installation_type = download ]; then
    rm -r lua
    rm release.zip

    if [ $version = "latest" ]; then
        url=https://github.com/${github_repo}/releases/latest/download/${os}-${neovim_version}.zip
    else
        url=https://github.com/${github_repo}/releases/download/${version}/${os}-${neovim_version}.zip
    fi

    wget $url -O release.zip
    unzip release.zip
else
    cargo build --release
    mkdir lua
    cp target/release/lib${crate_name}.so lua/${module_name}.so
    cp target/release/lib${crate_name}.dylib lua/${module_name}.so
fi;
```

**Reply by yutkat** [reply]
@ModProg Is this question already solved?

**Reply by ModProg** [reply, marked as answer]
well, it was more a question if someone knows a better way than the relatively complex installation instruction of:

> ```lua
> {
>     'ModProg/test.nvim',
>     run = function(plugin)
>         local await = require 'packer.async'.wait
>         local jobs = require 'packer.jobs'
> 
>         local neovim_version = vim.version().major .. '.' .. vim.version().minor
>         local os
>         if jit.os == "Linux" then
>             os = "linux"
>         elseif jit.os == "Windows" then
>             os = "windows"
>         else
>             os = "macos"
>         end
>         await(jobs.run({ "./install", "download", os, neovim_version, "latest" }, { cwd = plugin.install_path }))
>     end,
> }
> ```

I thought it would make sense to do the version selection inside neovim instead of relying on the output of `nvim --version`. But if packer doesn't have any other capability than cloneing for installing plugins, I'd say this can be closed.
